File: macos/Sources/LaTeXColabCore/FileTree.swift
import Foundation

/// One entry of the project file tree. `id` is the project-relative path.
public struct FileNode: Identifiable, Hashable {
    public let id: String
    public let name: String
    public let url: URL
    public let isDirectory: Bool
    public let children: [FileNode]?

    public init(id: String, name: String, url: URL, isDirectory: Bool, children: [FileNode]?) {
        self.id = id
        self.name = name
        self.url = url
        self.isDirectory = isDirectory
        self.children = children
    }

    public var relativePath: String { id }
    public var fileExtension: String { (name as NSString).pathExtension.lowercased() }
}

public enum FileTree {
    /// Directories that never show up in the tree or the zip.
    public static let excludedDirectories: Set<String> = [
        ".git", "__pycache__", ".venv", "node_modules", ".idea", ".vscode",
    ]
    /// LaTeX build artifacts hidden from the tree and left out of the zip.
    public static let excludedSuffixes: [String] = [
        ".aux", ".log", ".out", ".toc", ".fls", ".fdb_latexmk",
        ".synctex.gz", ".synctex", ".bbl", ".blg", ".nav", ".snm", ".vrb",
    ]
    public static let excludedFiles: Set<String> = [".DS_Store"]

    public static let textExtensions: Set<String> = [
        "tex", "bib", "cls", "sty", "md", "txt", "json", "yaml", "yml",
        "toml", "cfg", "ini", "gitignore", "sh", "py", "js", "ts", "html",
        "css", "csv", "tsv", "bst", "def", "clo", "ltx", "dtx",
    ]
    static let textFileNames: Set<String> = [".gitignore", ".gitattributes", "Makefile", "latexmkrc", ".latexmkrc"]

    public static func isExcluded(fileName: String) -> Bool {
        if excludedFiles.contains(fileName) { return true }
        return excludedSuffixes.contains { fileName.hasSuffix($0) }
    }

    public static func isTextFile(named name: String) -> Bool {
        if textFileNames.contains(name) { return true }
        return textExtensions.contains((name as NSString).pathExtension.lowercased())
    }

    public static func build(root: URL) -> [FileNode] {
        buildChildren(of: root, root: root)
    }

    private static func buildChildren(of directory: URL, root: URL) -> [FileNode] {
        let fm = FileManager.default
        guard let items = try? fm.contentsOfDirectory(
            at: directory,
            includingPropertiesForKeys: [.isDirectoryKey],
            options: []
        ) else { return [] }

        var nodes: [FileNode] = []
        for item in items {
            let isDir = (try? item.resourceValues(forKeys: [.isDirectoryKey]).isDirectory) ?? false
            let name = item.lastPathComponent
            let rel = relativePath(of: item, root: root)
            if isDir {
                if excludedDirectories.contains(name) { continue }
                nodes.append(FileNode(id: rel, name: name, url: item, isDirectory: true,
                                      children: buildChildren(of: item, root: root)))
            } else {
                if isExcluded(fileName: name) { continue }
                nodes.append(FileNode(id: rel, name: name, url: item, isDirectory: false, children: nil))
            }
        }
        nodes.sort { a, b in
            if a.isDirectory != b.isDirectory { return a.isDirectory }
            return a.name.localizedCaseInsensitiveCompare(b.name) == .orderedAscending
        }
        return nodes
    }

    public static func relativePath(of url: URL, root: URL) -> String {
        let rootPath = root.standardizedFileURL.path
        let path = url.standardizedFileURL.path
        if path == rootPath { return "" }
        if path.hasPrefix(rootPath + "/") {
            return String(path.dropFirst(rootPath.count + 1))
        }
        return url.lastPathComponent
    }

    /// Depth-first list of every file node (directories excluded).
    public static func flattenFiles(_ nodes: [FileNode]) -> [FileNode] {
        var out: [FileNode] = []
        for n in nodes {
            if n.isDirectory {
                out.append(contentsOf: flattenFiles(n.children ?? []))
            } else {
                out.append(n)
            }
        }
        return out
    }

    public static func find(_ path: String, in nodes: [FileNode]) -> FileNode? {
        for n in nodes {
            if n.id == path { return n }
            if let c = n.children, let hit = find(path, in: c) { return hit }
        }
        return nil
    }
}

File: linux/latexcolab/core/filetree.py
"""Project file tree (port of FileTree.swift)."""

from __future__ import annotations

import os
from dataclasses import dataclass
from pathlib import Path

#: Directories that never show up in the tree or the zip.
EXCLUDED_DIRECTORIES = {".git", "__pycache__", ".venv", "node_modules", ".idea", ".vscode"}

#: LaTeX build artifacts hidden from the tree and left out of the zip.
EXCLUDED_SUFFIXES = [
    ".aux", ".log", ".out", ".toc", ".fls", ".fdb_latexmk",
    ".synctex.gz", ".synctex", ".bbl", ".blg", ".nav", ".snm", ".vrb",
]

EXCLUDED_FILES = {".DS_Store"}

TEXT_EXTENSIONS = {
    "tex", "bib", "cls", "sty", "md", "txt", "json", "yaml", "yml",
    "toml", "cfg", "ini", "gitignore", "sh", "py", "js", "ts", "html",
    "css", "csv", "tsv", "bst", "def", "clo", "ltx", "dtx",
}

TEXT_FILE_NAMES = {".gitignore", ".gitattributes", "Makefile", "latexmkrc", ".latexmkrc"}


@dataclass(frozen=True)
class FileNode:
    id: str            # project-relative path
    name: str
    path: str          # absolute path
    is_directory: bool
    children: tuple["FileNode", ...] | None = None

    @property
    def relative_path(self) -> str:
        return self.id

    @property
    def file_extension(self) -> str:
        return os.path.splitext(self.name)[1].lstrip(".").lower()


def is_excluded(file_name: str) -> bool:
    if file_name in EXCLUDED_FILES:
        return True
    return any(file_name.endswith(s) for s in EXCLUDED_SUFFIXES)


def is_text_file(name: str) -> bool:
    if name in TEXT_FILE_NAMES:
        return True
    return os.path.splitext(name)[1].lstrip(".").lower() in TEXT_EXTENSIONS


def relative_path(path: str | os.PathLike, root: str | os.PathLike) -> str:
    root_path = os.path.normpath(str(root))
    p = os.path.normpath(str(path))
    if p == root_path:
        return ""
    if p.startswith(root_path + os.sep):
        return p[len(root_path) + 1:]
    return os.path.basename(p)


def build(root: str | os.PathLike) -> list[FileNode]:
    return _build_children(str(root), str(root))


def _build_children(directory: str, root: str) -> list[FileNode]:
    try:
        entries = list(os.scandir(directory))
    except OSError:
        return []

    nodes: list[FileNode] = []
    for entry in entries:
        name = entry.name
        try:
            is_dir = entry.is_dir(follow_symlinks=False)
        except OSError:
            continue
        rel = relative_path(entry.path, root)
        if is_dir:
            if name in EXCLUDED_DIRECTORIES:
                continue
            nodes.append(FileNode(rel, name, entry.path, True,
                                  tuple(_build_children(entry.path, root))))
        else:
            if is_excluded(name):
                continue
            nodes.append(FileNode(rel, name, entry.path, False, None))

    nodes.sort(key=lambda n: (not n.is_directory, n.name.lower(), n.name))
    return nodes


def flatten_files(nodes: list[FileNode] | tuple[FileNode, ...]) -> list[FileNode]:
    """Depth-first list of every file node (directories excluded)."""
    out: list[FileNode] = []
    for n in nodes:
        if n.is_directory:
            out.extend(flatten_files(n.children or ()))
        else:
            out.append(n)
    return out


def find(path: str, nodes: list[FileNode] | tuple[FileNode, ...]) -> FileNode | None:
    for n in nodes:
        if n.id == path:
            return n
        if n.children:
            hit = find(path, n.children)
            if hit is not None:
                return hit
    return None


def read_text(path: str | os.PathLike) -> str | None:
    """Read a UTF-8 text file, normalising CRLF. None when it is not readable text."""
    try:
        data = Path(path).read_bytes()
    except OSError:
        return None
    try:
        return data.decode("utf-8").replace("\r\n", "\n")
    except UnicodeDecodeError:
        return None
